wdl-format/src/format/comments.rs

```diff
@@ -0,0 +1,153 @@
+use wdl_ast::Comment;
+use wdl_ast::SyntaxElement;
+use wdl_ast::SyntaxKind;
+
+use super::FormatState;
+use super::Formattable;
+use super::NEWLINE;
+
+/// Inline comment space constant used for formatting.
+pub const INLINE_COMMENT_SPACE: &str = "  ";
+
+impl Formattable for Comment {
+    fn format(&self, buffer: &mut String, _state: &mut FormatState) -> Result<()> {
+        let comment = self.as_str().trim();
+        write!(buffer, "{}{}", comment, NEWLINE)?;
+        Ok(())
+    }
+
+    fn syntax_element(&self) -> SyntaxElement {
+        SyntaxElement::Token(self.syntax().clone())
+    }
+}
+
+/// Format comments that preceed a node.
+pub fn format_preceding_comments(
+    element: &SyntaxElement,
+    buffer: &mut String,
+    state: &mut FormatState,
+    would_be_interrupting: bool,
+) -> Result<()> {
+    // This walks _backwards_ through the syntax tree to find comments
+    // so we must collect them in a vector and later reverse them to get them in the
+    // correct order.
+    let mut preceding_comments = Vec::new();
+
+    let mut prev = element.prev_sibling_or_token();
+    while let Some(cur) = prev {
+        match cur.kind() {
+            SyntaxKind::Comment => {
+                // Ensure this comment "belongs" to the root element.
+                // A preceding comment on a blank line is considered to belong to the element.
+                // Othewise, the comment "belongs" to whatever
+                // else is on that line.
+                if let Some(before_cur) = cur.prev_sibling_or_token() {
+                    match before_cur.kind() {
+                        SyntaxKind::Whitespace => {
+                            if before_cur.to_string().contains('\n') {
+                                // The 'cur' comment is on is on its own line.
+                                // It "belongs" to the current element.
+                                let comment = Comment::cast(
+                                    cur.as_token().expect("Comment should be a token").clone(),
+                                )
+                                .expect("Comment should cast to a comment");
+                                preceding_comments.push(comment);
+                            }
+                        }
+                        _ => {
+                            // The 'cur' comment is on the same line as this
+                            // token. It "belongs"
+                            // to whatever is currently being processed.
+                        }
+                    }
+                }
+            }
+            SyntaxKind::Whitespace => {
+                // Ignore
+            }
+            _ => {
+                // We've backed up to non-trivia, so we can stop
+                break;
+            }
+        }
+        prev = cur.prev_sibling_or_token()
+    }
+
+    let comments_found = !preceding_comments.is_empty();
+    if comments_found && would_be_interrupting && !state.interrupted() {
```

Comment:
Ah, I was confused by this for a moment, but this is an example of an interrupting, preceding comment.
```
version
# a comment
1.1
```